File: examples/quickstart/java/MyServer.java
package com.example.myapp;

import com.example.myapp.checkout.*;
import com.bitmechanic.pulserpc.*;
import java.util.*;

public class MyServer {
    static List<Product> products = new ArrayList<>();

    static {
        Product p1 = new Product();
        p1.setProductId("prod001");
        p1.setName("Wireless Mouse");
        p1.setDescription("Ergonomic mouse");
        p1.setPrice(29.99);
        p1.setStock(50);
        p1.setImageUrl("https://example.com/mouse.jpg");
        products.add(p1);

        Product p2 = new Product();
        p2.setProductId("prod002");
        p2.setName("Mechanical Keyboard");
        p2.setDescription("RGB keyboard");
        p2.setPrice(89.99);
        p2.setStock(25);
        p2.setImageUrl("https://example.com/keyboard.jpg");
        products.add(p2);
    }

    static Map<String, Cart> carts = new HashMap<>();
    static Map<String, Order> orders = new HashMap<>();

    static class CatalogServiceImpl implements CatalogService {
        public List<Product> listProducts() {
            return products;
        }

        public Product getProduct(String productId) {
            return products.stream()
                .filter(p -> p.getProductId().equals(productId))
                .findFirst().orElse(null);
        }
    }

    static class CartServiceImpl implements CartService {
        public Cart addToCart(AddToCartRequest request) {
            String cartId = request.getCartId();
            if (cartId == null || cartId.isEmpty()) {
                cartId = "cart_" + (int)(Math.random() * 9000 + 1000);
            }

            Cart cart = carts.get(cartId);
            if (cart == null) {
                cart = new Cart();
                cart.setCartId(cartId);
                cart.setItems(new ArrayList<>());
                cart.setSubtotal(0.0);
                carts.put(cartId, cart);
            }

            Product product = products.stream()
                .filter(p -> p.getProductId().equals(request.getProductId()))
                .findFirst().orElseThrow(() -> new RPCError(-32602, "Product not found"));

            CartItem item = new CartItem();
            item.setProductId(request.getProductId());
            item.setQuantity(request.getQuantity());
            item.setPrice(product.getPrice());
            cart.getItems().add(item);
            cart.setSubtotal(cart.getItems().stream().mapToDouble(i -> i.getPrice() * i.getQuantity()).sum());

            return cart;
        }

        public Cart getCart(String cartId) {
            return carts.get(cartId);
        }

        public boolean clearCart(String cartId) {
            Cart cart = carts.get(cartId);
            if (cart != null) {
                cart.getItems().clear();
                cart.setSubtotal(0.0);
                return true;
            }
            return false;
        }
    }

    static class OrderServiceImpl implements OrderService {
        public CheckoutResponse createOrder(CreateOrderRequest request) {
            Cart cart = carts.get(request.getCartId());
            if (cart == null) {
                throw new RPCError(1001, "CartNotFound: Cart does not exist");
            }

            if (cart.getItems().isEmpty()) {
                throw new RPCError(1002, "CartEmpty: Cannot create order from empty cart");
            }

            String orderId = "order_" + (int)(Math.random() * 90000 + 10000);
            Order order = new Order();
            order.setOrderId(orderId);
            order.setCart(cart);
            order.setShippingAddress(request.getShippingAddress());
            order.setPaymentMethod(request.getPaymentMethod());
            order.setStatus(OrderStatus.pending);
            order.setTotal(cart.getSubtotal());
            order.setCreatedAt((int)(System.currentTimeMillis() / 1000));
            orders.put(orderId, order);

            CheckoutResponse resp = new CheckoutResponse();
            resp.setOrderId(orderId);
            resp.setMessage("Order created successfully");
            return resp;
        }

        public Order getOrder(String orderId) {
            return orders.get(orderId);
        }
    }

    public static void main(String[] args) throws Exception {
        JsonParser jsonParser = new JacksonJsonParser();
        Server server = new Server(8080, jsonParser);
        server.register("CatalogService", new CatalogServiceImpl());
        server.register("CartService", new CartServiceImpl());
        server.register("OrderService", new OrderServiceImpl());
        server.start();
    }
}

File: examples/quickstart/csharp/TestServer/MyServer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using PulseRPC;
using checkout;

public class CatalogServiceImpl : ICatalogService
{
    private static readonly List<Product> Products = new List<Product>
    {
        new Product { ProductId = "prod001", Name = "Wireless Mouse", Description = "Ergonomic mouse", Price = 29.99, Stock = 50, ImageUrl = "https://example.com/mouse.jpg" },
        new Product { ProductId = "prod002", Name = "Mechanical Keyboard", Description = "RGB keyboard", Price = 89.99, Stock = 25, ImageUrl = "https://example.com/keyboard.jpg" }
    };

    public List<Product> listProducts()
    {
        return Products;
    }

    public Product? getProduct(string productId)
    {
        var product = Products.FirstOrDefault(p => p.ProductId == productId);
        if (product == null)
            throw new RPCError(-32602, "Product not found");
        return product;
    }

}

public class CartServiceImpl : ICartService
{
    internal readonly Dictionary<string, Cart> _carts = new Dictionary<string, Cart>();
    private readonly CatalogServiceImpl _catalogService;

    public CartServiceImpl(CatalogServiceImpl catalogService)
    {
        _catalogService = catalogService;
    }

    public Cart addToCart(AddToCartRequest request)
    {
        var cartId = request.CartId ?? $"cart_{new Random().Next(1000, 9999)}";

        if (!_carts.TryGetValue(cartId, out var cart))
        {
            cart = new Cart { CartId = cartId, Items = new List<CartItem>(), Subtotal = 0 };
            _carts[cartId] = cart;
        }

        var product = _catalogService.listProducts().FirstOrDefault(p => p.ProductId == request.ProductId);
        if (product == null)
            throw new RPCError(-32602, "Product not found");

        cart.Items.Add(new CartItem { ProductId = request.ProductId, Quantity = request.Quantity, Price = (double)product.Price });
        cart.Subtotal = (double)cart.Items.Sum(i => i.Price * i.Quantity);

        return cart;
    }

    public Cart? getCart(string cartId)
    {
        return _carts.TryGetValue(cartId, out var cart) ? cart : null;
    }

    public bool clearCart(string cartId)
    {
        if (_carts.TryGetValue(cartId, out var cart))
        {
            cart.Items.Clear();
            cart.Subtotal = 0;
            return true;
        }
        return false;
    }
}

class OrderServiceImpl : IOrderService
{
    private readonly Dictionary<string, Cart> _carts;
    private readonly Dictionary<string, Order> _orders = new Dictionary<string, Order>();

    public OrderServiceImpl(Dictionary<string, Cart> carts)
    {
        _carts = carts;
    }

    public CheckoutResponse createOrder(CreateOrderRequest request)
    {
        if (!_carts.TryGetValue(request.CartId, out var cart))
            throw new RPCError(1001, "CartNotFound: Cart does not exist");

        if (cart.Items.Count == 0)
            throw new RPCError(1002, "CartEmpty: Cannot create order from empty cart");

        var orderId = $"order_{new Random().Next(10000, 99999)}";
        var order = new Order
        {
            OrderId = orderId,
            Cart = cart,
            ShippingAddress = request.ShippingAddress,
            PaymentMethod = request.PaymentMethod,
            Status = OrderStatus.pending,
            Total = (double)cart.Subtotal,
            CreatedAt = (int)DateTimeOffset.UtcNow.ToUnixTimeSeconds()
        };

        _orders[orderId] = order;
        return new CheckoutResponse { OrderId = orderId };
    }

    public Order? getOrder(string orderId)
    {
        return _orders.TryGetValue(orderId, out var order) ? order : null;
    }
}

class Program
{
    static async Task Main(string[] args)
    {
        var server = new PulseRPCServer();
        var catalogService = new CatalogServiceImpl();
        var cartService = new CartServiceImpl(catalogService);

        server.RegisterCatalogService(catalogService);
        server.RegisterCartService(cartService);
        server.RegisterOrderService(new OrderServiceImpl(cartService._carts));

        Console.WriteLine("Server starting on http://localhost:8080");
        await server.RunAsync("0.0.0.0", 8080);
    }
}
